sdc11073/wsdiscovery.py

```diff
@@ -820,45 +831,42 @@ def removeSourceAddr(self, addr):
     def addUnicastMessage(self, env, addr, port, initialDelay=0):
         msg = Message(env, addr, port, Message.UNICAST, initialDelay)
         self._logger.debug('addUnicastMessage: adding message Id %s. delay=%.2f'.format(env.getMessageId(), initialDelay))
-        self._queue.put(msg)
+        self._repeated_enqueue_msg(msg, initialDelay, UNICAST_UDP_REPEAT, UNICAST_UDP_MIN_DELAY,
+                                   UNICAST_UDP_MAX_DELAY, UNICAST_UDP_UPPER_DELAY)
 
     def addMulticastMessage(self, env, addr, port, initialDelay=0):
         msg = Message(env, addr, port, Message.MULTICAST, initialDelay)
         self._logger.debug('addMulticastMessage: adding message Id %s. delay=%.2f'.format(env.getMessageId(), initialDelay))
-        self._queue.put(msg)
+        self._repeated_enqueue_msg(msg, initialDelay, MULTICAST_UDP_REPEAT, MULTICAST_UDP_MIN_DELAY,
+                                   MULTICAST_UDP_MAX_DELAY, MULTICAST_UDP_UPPER_DELAY)
+
+    def _repeated_enqueue_msg(self, msg, initial_delay_ms, repeat, min_delay_ms, max_delay_ms, upper_delay_ms):
+        next_send = time.time() + initial_delay_ms/1000.0
+        dt = (min_delay_ms + ((max_delay_ms - min_delay_ms) * random.random())) / 2000.0
+        self._send_queue.put(self._EnqueuedMessage(next_send, msg))
+        for _ in range(repeat):
+            next_send += dt
+            self._send_queue.put(self._EnqueuedMessage(next_send, msg))
+            dt = min(dt*2, upper_delay_ms)
 
     def _run_send(self):
         ''' run by thread'''
-        while not self._quitSendEvent.is_set():  # or self._queue:
-            if not self._queue.empty():
-                try:
-                    sz = self._queue.qsize()
-                    if sz > 800:
-                        self._logger.error('_queue size =%d', sz)
-                    elif sz > 500:
-                        self._logger.warn('_queue size =%d', sz)
-                    elif sz > 100:
-                        self._logger.info('_queue size =%d', sz)
-                    for dummy in range(self._queue.qsize()):
-                        msg = self._queue.get()
-                        if msg.canSend():
-                            self._sendMsg(msg)
-                            msg.refresh()
-                            if not msg.isFinished():
-                                self._queue.put(msg)
-                        else:
-                            self._queue.put(msg)
-                except:
-                    self._logger.error('_run_send:{}'.format(traceback.format_exc()))
-                time.sleep(0.02)
+        while not self._quitSendEvent.is_set():
+            try:
+                enqueued_msg = self._send_queue.get(timeout=0.1)
+            except queue.Empty:
+                pass
             else:
-                time.sleep(0.2)
+                now = time.time()
+                if enqueued_msg.send_time > now:

```

Comment:
So what happens when a new message is added to the queue which needs to be send before this one? If this message is way way more delayed, it blocks other more urgent messages which have been added to the queue later on.

Reply:
I think we can live with that. In the end this is not about priority, but only about ramdomness. The charme of this is its simplicity.